[SDL 0317] SDL Protocol Security Specification (#41)
smartdevicelink/protocol_spec: README.md

      <b>Frame Type = 0x00 (Control Frame)</b><br>
      0x0 - 0xFFFFFFFF reserved.<br>
      <b>Frame Type = 0x02 (First Frame)</b><br>

>       0x08 The data size for a first frame is always 8 bytes. In the payload, the first four bytes denote the Total Size of the data contained in all consecutive frames, and the second four bytes denote the number of consecutive frames following this one<br>

Reviewer: > 0x08 The data size for a first frame is always 8 bytes.

Any reason for removing this note?

Author: @jacobkeeler looks like it was removed by mistake. I restored the first sentence here. https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

>       The data size for a first frame is always 8 bytes. In the payload, the first four bytes denote the Total Size of the data contained in all consecutive frames. This is always the size of whole non-encrypted payload (even if consecutive frames are encrypted). The second four bytes denote the number of consecutive frames following this one<br>

Reviewer: ```suggestion
      0x08 The data size for a first frame is always 8 bytes. In the payload, the first four bytes denote the Total Size of the data contained in all consecutive frames. This is always the size of whole non-encrypted payload (even if consecutive frames are encrypted). The second four bytes denote the number of consecutive frames following this one<br>
```

Author: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/a323ff3e488b4710003d7e905dde521ac4a5e2f5

### 5.1 Control Service
>Required: All Protocol Versions

The control service is the lowest level service available. While Control Frame packets are used frequently, the control service itself is rarely used.

#### 5.1.1 Security Query

> The TLS Payload is sent in a binary query header. The size is 12 bytes which matches the size of the RPC Payload Binary Header.

Reviewer: ```suggestion
When establishing a secure connection, the TLS Payload is sent in a control service message with a binary query header. The size of this header is 12 bytes, similar to the RPC Payload Binary Header.
```

Author: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

##### 5.1.1.1 Payload

> The security query is able to contain JSON data as well as binary data. During the handshake the TLS handshake data is sent as binary data. In case of an error, a notification is sent with an error code and error message as JSON data. See "Error handling" section for details.

Reviewer: > During the handshake the TLS handshake data is sent as binary data.

I think we should probably include a subsection `5.x.x` around here describing this, even if it just links out to the "Secured Connection" section:

```

**Request:**

<table>
  <tr>
    <th align="center">Binary Data</th>
  </tr>
  <tr>
    <td align="center" width="800">SSL Handshake Request</td>
  </tr>
</table>

**Response:**

<table>
  <tr>
    <th align="center">Binary Data</th>
  </tr>
  <tr>
    <td align="center" width="800">SSL Handshake Response</td>
  </tr>
</table>
```

This might require renumbering a few of the sections

Author: @jacobkeeler moved into a separate section "Send Handshake Data"
See https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

  <tr><td align="center">Binary Data</td></tr>
</table>

##### 5.1.1.2 Binary Header

<table width="100%">
  <tr>
    <th width="25%">Byte 1</th>
    <th width="25%">Byte 2</th>
    <th width="25%">Byte 3</th>
    <th width="25%">Byte 4</th>
  </tr>
  <tr>
    <td>Query Type</td>
    <td colspan="3" align="center">Query ID</td>
  </tr>
  <tr>
    <td colspan="4" align="center">Sequential Number</td>
  </tr>
  <tr>
    <td colspan="4" align="center">JSON Size</td>
  </tr>
</table>

###### 5.1.1.2.1 Binary Header Fields

<table width="100%">
  <tr>
    <th>Field</th>
    <th>Size</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>Query Type</td>
    <td>8 bit</td>
    <td>
      0x00 Request <br>
      0x01 - 0x0F Reserved<br>
      0x10 Response<br>
      0x11 - 0x1F Reserved<br>
      0x20 Notification<br>
      0x21 - 0xFE Reserved<br>
      0xFF Invalid Query Type
    </td>
  </tr>
  <tr>
    <td>Query ID</td>
    <td>24 bit</td>
    <td>
      0x000001 Send Handshake Data<br>
      0x000002 Send Internal Error<br>
      0x000003 - 0xFFFFFE Reserved<br>
      0xFFFFFF Invalid Query ID
    </td>
  </tr>
  <tr>
    <td>Sequential Number</td>
    <td>32 bit</td>
    <td>

>       Message ID can be set by the Mobile libraries to track security messages.

Reviewer: ```suggestion
      Message ID can be set by the mobile libraries to track security messages.<br>
```

Author: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

>       The system uses the same Message ID when replying to the query allowing the mobile to correlate messages.

Reviewer: ```suggestion
      The system uses the same Message ID when replying to the query allowing the mobile libraries to correlate messages.
```

Author: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

    </td>
  </tr>
  <tr>
    <td>JSON Size</td>
    <td>32 bit</td>

>     <td>The size of the JSON data following the binary query header.</td>

Reviewer: ```suggestion
    <td>
      The size of the JSON data following the binary query header.<br>
      Any additional data following the JSON data in the payload is binary data.
    </td>
```

Author: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

  </tr>
</table>

> #### 5.1.2 Error Frames

Reviewer: ```suggestion
#### 5.1.1.2 Send Internal Error
```
This should probably be a subsection of "Security Query", as should `Send Handshake Data`

Author: @jacobkeeler done. Now both "Send Handshake Data" and "Send Internal Error" are specified as separate subsections. https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

    <td>Unknown Query ID</td>
  </tr>
  <tr>
    <td>ERROR_NOT_SUPPORTED</td>
    <td>0x03</td><td>3</td>
    <td>SDL does not support encryption</td>
  </tr>
  <tr>
    <td>ERROR_SERVICE_ALREADY_PROTECTED</td>
    <td>0x04</td><td>4</td>
    <td>Received request to protect a service that was protected before</td>
  </tr>
  <tr>
    <td>ERROR_SERVICE_NOT_PROTECTED</td>
    <td>0x05</td><td>5</td>
    <td>Received handshake or encrypted data for not protected service</td>
  </tr>
  <tr>
    <td>ERROR_DECRYPTION_FAILED</td>
    <td>0x06</td><td>6</td>
    <td>Decryption failed</td>
  </tr>
  <tr>
    <td>ERROR_ENCRYPTION_FAILED</td>
    <td>0x07</td><td>7</td>
    <td>Encryption failed</td>
  </tr>
  <tr>
    <td>ERROR_SSL_INVALID_DATA</td>
    <td>0x08</td><td>8</td>
    <td>SSL invalid data</td>
  </tr>
  <tr>
    <td>ERROR_HANDSHAKE_FAILED</td>
    <td>0x09</td><td>9</td>
    <td>In case of all other handshake errors</td>
  </tr>
  <tr>
    <td>INVALID_CERT</td>
    <td>0x0A</td><td>10</td>
    <td>Handshake failed because certificate is invalid</td>
  </tr>
  <tr>
    <td>EXPIRED_CERT</td>
    <td>0x0B</td><td>11</td>
    <td>Handshake failed because certificate is expired</td>

>   </tr>

Reviewer: These 3 values are not defined in Core, where are these coming from?

Author: @jacobkeeler looks like it is the values from older SYNC generations that are not used in newer systems. I removed these three to avoid changes in SDL core. https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

### 5.1 Control Service
>Required: All Protocol Versions

The control service is the lowest level service available. While Control Frame packets are used frequently, the control service itself is rarely used.

#### 5.1.1 Security Query

When establishing a secure connection, the TLS Payload is sent in a control service message with a binary query header. The size of this header is 12 bytes, similar to the RPC Payload Binary Header.

##### 5.1.1.1 Payload

The security query is able to contain JSON data as well as binary data. During the handshake the TLS handshake data is sent as binary data. See "Send Handshake Data" section for details.

> In case of an error, a notification is sent with an error code and error message as JSON data. See "Error handling" section for details.

Reviewer: ```suggestion
In case of an error, a notification is sent with an error code and error message as JSON data. See "Send Internal Error" section for details.
```

Author: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/efddba1d5059d7aa763a2502acb600a726535001

  </tr>
</table>

###### 5.1.1.2.1 Binary Header Fields

<table width="100%">
  <tr>
    <th>Field</th>
    <th>Size</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>Query Type</td>
    <td>8 bit</td>
    <td>
      0x00 Request <br>
      0x01 - 0x0F Reserved<br>
      0x10 Response<br>
      0x11 - 0x1F Reserved<br>
      0x20 Notification<br>
      0x21 - 0xFE Reserved<br>
      0xFF Invalid Query Type
    </td>
  </tr>
  <tr>
    <td>Query ID</td>
    <td>24 bit</td>
    <td>
      0x000001 Send Handshake Data<br>
      0x000002 Send Internal Error<br>
      0x000003 - 0xFFFFFE Reserved<br>
      0xFFFFFF Invalid Query ID
    </td>
  </tr>
  <tr>
    <td>Sequential Number</td>
    <td>32 bit</td>
    <td>
      Message ID can be set by the mobile libraries to track security messages.<br>
      The system uses the same Message ID when replying to the query allowing the mobile libraries to correlate messages.
    </td>
  </tr>
  <tr>
    <td>JSON Size</td>
    <td>32 bit</td>
    <td>
      The size of the JSON data following the binary query header.<br>
      Any additional data following the JSON data in the payload is binary data.
    </td>
  </tr>
</table>

> ##### 5.1.1.3 Send Handshake Data
> 

Reviewer: ```suggestion
##### 5.1.1.3 Send Handshake Data

When performing the TLS handshake, the server sends a "Send Handshake Data" request containing its handshake data to the client, and the client sends a response with its own handshake data accordingly.

```
Adding description for this message.

Author: @jacobkeeler added in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/166cb9cdfbcf6d97e9d3e38854426ae744f14b7d

> **Request:**

Reviewer: ```suggestion
##### 5.1.1.2.1 Request Payload

**Note:** Prior to SDL Core version 8.0.0, Core sent the "Notification" type (`0x20`) in this message instead of "Request" (`0x00`). Client libraries should account for this when communicating with older versions of SDL Core.

```
Updating subsection names for organization and added note about previous Core behavior

Author: @jacobkeeler applied. I also used `5.1.1.3.1` section number as from my understanding this is a sub section of `5.1.1.3 Send Handshake Data`
See https://github.com/smartdevicelink/protocol_spec/pull/41/commits/166cb9cdfbcf6d97e9d3e38854426ae744f14b7d

<table width="100%">
  <tr>
    <th align="center">Binary Data</th>
  </tr>
  <tr>
    <td align="center">SSL Handshake Request</td>
  </tr>

> </table>

Reviewer: ```suggestion
<table width="100%">
  <tr>
    <th colspan="4">Binary Query Header</th>
  <tr>
  <tr>
    <th>Query Type</th>
    <th>TLS Message Type</th>
    <th>Sequential Number</th>
    <th>JSON Size</th>
  </tr>
  <tr>
    <td>Request</td>
    <td>Send Handshake Data</td>
    <td>Any number to be used to correlate query messages</td>
    <td>Zero</td>
  </tr>
  <tr>
    <td>0x00</td>
    <td>0x000001</td>
    <td>0xNNNNNNNN</td>
    <td>0x00000000</td>
  </tr>
  <tr>
    <th align="center">Binary Data: SSL Handshake Request</td>
  </tr>
</table>
```
Should use the more detailed definitions used in the proposal for clarity

Author: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/166cb9cdfbcf6d97e9d3e38854426ae744f14b7d

> **Response:**

Reviewer: ```suggestion
##### 5.1.1.2.2 Response Payload

```

Author: @jacobkeeler applied and updated section numbering
See https://github.com/smartdevicelink/protocol_spec/pull/41/commits/166cb9cdfbcf6d97e9d3e38854426ae744f14b7d

<table width="100%">
  <tr>
    <th align="center">Binary Data</th>
  </tr>
  <tr>
    <td align="center">SSL Handshake Response</td>
  </tr>

> </table>

Reviewer: ```suggestion
<table width="100%">
  <tr>
    <th colspan="4">Binary Query Header</th>
  <tr>
  <tr>
    <th>Query Type</th>
    <th>TLS Message Type</th>
    <th>Sequential Number</th>
    <th>JSON Size</th>
  </tr>
  <tr>
    <td>Response</td>
    <td>Send Handshake Data</td>
    <td>Any number to be used to correlate query messages</td>
    <td>Zero</td>
  </tr>
  <tr>
    <td>0x10</td>
    <td>0x000001</td>
    <td>0xNNNNNNNN</td>
    <td>0x00000000</td>
  </tr>
  <tr>
    <th align="center">Binary Data: SSL Handshake Response</td>
  </tr>
</table>
```
Should use the more detailed definitions used in the proposal for clarity

Author: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/166cb9cdfbcf6d97e9d3e38854426ae744f14b7d

#### 5.1.1.4 Send Internal Error

If an error occurs during the TLS handshake, a notification is sent with both JSON data and binary data describing the error. The JSON data contains the error code and an error text. The binary data is one single byte and only contains the error code.

The error code in JSON data and the binary data are the same value from the same code list.

##### 5.1.1.4.1 Payload

The following query header is used by the system and the application to send error messages.

<table width="100%">
  <tr>
    <th colspan="4">Binary Query Header</th>
  <tr>
  <tr>
    <th>Query Type</th>
    <th>TLS Message Type</th>
    <th>Sequential Number</th>
    <th>JSON Size</th>
  </tr>
  <tr>
    <td>Notification</td>
    <td>Send Internal Error</td>
    <td>Any number to be used to correlate query messages</td>
    <td>Size of the JSON data</td>
  </tr>
  <tr>
    <td>0x20</td>
    <td>0x000002</td>
    <td>0xNNNNNNNN</td>
    <td>0xNNNNNNNN</td>

>   </tr>

Reviewer: ```suggestion
  <tr>
    <td>Notification</td>
    <td>Send Internal Error</td>
    <td>Unused</td>
    <td>Size of the JSON data</td>
  </tr>
  <tr>
    <td>0x20</td>
    <td>0x000002</td>
    <td>0xNNNNNNNN</td>
    <td>0xNNNNNNNN</td>
  </tr>
```

Author: @jacobkeeler fixed in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/166cb9cdfbcf6d97e9d3e38854426ae744f14b7d

    <td>JSON Size</td>
    <td>32 bit</td>
    <td>
      The size of the JSON data following the binary query header.<br>
      Any additional data following the JSON data in the payload is binary data.
    </td>
  </tr>
</table>

##### 5.1.1.3 Send Handshake Data

When performing the TLS handshake, the server sends a "Send Handshake Data" request containing its handshake data to the client, and the client sends a response with its own handshake data accordingly.

###### 5.1.1.3.1 Request Payload

**Note:** Prior to SDL Core version 8.0.0, Core sent the "Notification" type (`0x20`) in this message instead of "Request" (`0x00`). Client libraries should account for this when communicating with older versions of SDL Core.

<table width="100%">
  <tr>
    <th colspan="4">Binary Query Header</th>
  <tr>
  <tr>
    <th>Query Type</th>
    <th>TLS Message Type</th>
    <th>Sequential Number</th>
    <th>JSON Size</th>
  </tr>
  <tr>
    <td>Request</td>
    <td>Send Handshake Data</td>
    <td>Any number to be used to correlate query messages</td>
    <td>Zero</td>
  </tr>
  <tr>
    <td>0x00</td>
    <td>0x000001</td>
    <td>0xNNNNNNNN</td>
    <td>0x00000000</td>
  </tr>
  <tr>

>     <th align="center">Binary Data: SSL Handshake Request</td>

Reviewer: ```suggestion
    <th align="center" colspan="4">Binary Data: SSL Handshake Request</td>
```

Author: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/26e0db45251d3f8a3667b2f66e38660467ac001c

  </tr>
</table>

###### 5.1.1.3.2 Response Payload
<table width="100%">
  <tr>
    <th colspan="4">Binary Query Header</th>
  <tr>
  <tr>
    <th>Query Type</th>
    <th>TLS Message Type</th>
    <th>Sequential Number</th>
    <th>JSON Size</th>
  </tr>
  <tr>
    <td>Response</td>
    <td>Send Handshake Data</td>
    <td>Any number to be used to correlate query messages</td>
    <td>Zero</td>
  </tr>
  <tr>
    <td>0x10</td>
    <td>0x000001</td>
    <td>0xNNNNNNNN</td>
    <td>0x00000000</td>
  </tr>
  <tr>

>     <th align="center">Binary Data: SSL Handshake Response</td>

Reviewer: ```suggestion
    <th align="center" colspan="4">Binary Data: SSL Handshake Response</td>
```

Author: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/26e0db45251d3f8a3667b2f66e38660467ac001c

    <th>Byte</th>
    <th>Value</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>ERROR_SUCCESS</td>
    <td>0x00</td><td>0</td>
    <td>Internal Security Manager value</td>
  </tr>
  <tr>
    <td>ERROR_INVALID_QUERY_SIZE</td>
    <td>0x01</td><td>1</td>
    <td>Wrong size of query data</td>
  </tr>
  <tr>
    <td>ERROR_INVALID_QUERY_ID</td>
    <td>0x02</td><td>2</td>
    <td>Unknown Query ID</td>
  </tr>
  <tr>
    <td>ERROR_NOT_SUPPORTED</td>
    <td>0x03</td><td>3</td>
    <td>SDL does not support encryption</td>
  </tr>
  <tr>
    <td>ERROR_SERVICE_ALREADY_PROTECTED</td>
    <td>0x04</td><td>4</td>
    <td>Received request to protect a service that was protected before</td>
  </tr>
  <tr>
    <td>ERROR_SERVICE_NOT_PROTECTED</td>
    <td>0x05</td><td>5</td>
    <td>Received handshake or encrypted data for not protected service</td>
  </tr>
  <tr>
    <td>ERROR_DECRYPTION_FAILED</td>
    <td>0x06</td><td>6</td>
    <td>Decryption failed</td>
  </tr>
  <tr>
    <td>ERROR_ENCRYPTION_FAILED</td>
    <td>0x07</td><td>7</td>
    <td>Encryption failed</td>
  </tr>
  <tr>
    <td>ERROR_SSL_INVALID_DATA</td>
    <td>0x08</td><td>8</td>
    <td>SSL invalid data</td>

>   </tr>

Reviewer: ```suggestion
  </tr>
  <tr>
    <td>ERROR_HANDSHAKE_FAILED</td>
    <td>0x09</td><td>9</td>
    <td>In case of all other handshake errors</td>
  </tr>
  <tr>
    <td>INVALID_CERT</td>
    <td>0x0A</td><td>10</td>
    <td>Handshake failed because certificate is invalid</td>
  </tr>
  <tr>
    <td>EXPIRED_CERT</td>
    <td>0x0B</td><td>11</td>
    <td>Handshake failed because certificate is expired</td>
  </tr>
```
Seems that these are included in the proposal, so they need to be in this PR, they've been implemented in Core respectively: https://github.com/smartdevicelink/sdl_core/pull/3773

Author: @jacobkeeler applied in 347cbb864affdcf3f4e5afd1623b845b710a5c67
Should I rename `INVALID_CERT` -> `ERROR_INVALID_CERT` and `EXPIRED_CERT` -> `ERROR_EXPIRED_CERT` like in PR?

## 7. Secured Communication

It is possible to establish a secured and encrypted communication with the system by setting the frame header encryption flag to `1` when starting a new service or by sending another `StartService` with the encryption flag set to `1` when the service is already established (this the required flow for the RPC service). If the authentication is successful, the system will reply with a `StartService ACK` frame with the encryption flag also set to `1` indicating that encrypted data is now accepted. If the authentication fails for some reason, the system will reset the TLS connection and return a `StartService NAK` frame.

Before the encryption of RPC service is enabled (encryption is not available), SDL Core rejects any RPC request with result code `ENCRYPTION_NEEDED` if the RPC needs protection (please see policy updates for which RPCs need protection). SDL Core continues processing an RPC request if the RPC does not need protection. SDL Core sends a notification only if the notification RPC does not need protection.

> After the encryption of RPC service is enabled (encryption is available), SDL Core rejects any unencrypted RPC requests with result code `ENCRYPTION_NEEDED` with the unencrypted response if the RPC needs protection. SDL Core continues processing an unencrypted RPC request if the RPC does not need protection and responds with an unencrypted response. SDL Core continues processing an encrypted RPC request if the RPC needs protection and responds with an encrypted response. SDL Core sends an unencrypted notification if the RPC does not need protection. SDL Core sends an encrypted notification if the RPC needs protection. In addition, SDL Core shall continue processing an encrypted RPC request if the RPC does not need protection and responds with an encrypted response.

Reviewer: This is quite a cluttered section which describes a lot of different scenarios quickly, perhaps it would be best to lay out these scenarios in table format?

Author: @jacobkeeler I converted these 2 paragraphs into a table format. Hopefully, now it's easier to read. https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

> ### 7.1 Authentication

Reviewer: I'm wondering whether the protocol spec is the best place for all of this info in `7.1`, it might be better to include this in the SDL Overview Guides and link out to it here. It has several references to app and Core configurations that probably aren't really relevant to the protocol

It might still be worth including the TLS handshake diagram here, but most of the other information I feel could be moved to a better place.

Author: @jacobkeeler could you advise please if [this](https://github.com/smartdevicelink/sdl_core_guides/tree/master/docs/Feature%20Documentation) is the right place for moving out the information from this section?

Author: @jacobkeeler updated in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/efddba1d5059d7aa763a2502acb600a726535001

Reviewer: @AKalinich-Luxoft I think the best place for this information would be the SDL Overview Guides, since it relates to both the mobile libraries and SDL Core. It might need to be in its own "Protected Services" page for now, as the existing "RPC Encryption" guide doesn't really cover encrypted video and audio services.

Author: @jacobkeeler I added a link to SDL Overview Guides in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/b25ef60451a378bf1a7adf7701eaa96481e93454
Also, I opened [PR](https://github.com/smartdevicelink/sdl_overview_guides/pull/80) containing moved changes. The link in protocol_spec will become valid once this PR is merged. Could you please review it as well?

The below diagram shows the sequence of how the TLS handshake exchanges certificates to compute the master secret.

![TLS Handshake activity diagram](https://user-images.githubusercontent.com/5848997/122258220-cb8b7100-ce9e-11eb-9b2a-a6194b0d1b68.png)

> The authentication is done using TLS handshake. The TLS handshake process is defined by TLS and is not part of the SDL protocol. The handshake is designed as a client server communication which is configurable in the system settings. An application must take the role of a server where the system is the client. The client entity will initiate a TLS handshake with the corresponding security manager of the server. The client will do this only if the server was not authenticated before in the current transport connection. According to the TLS handshake process the peer certificate can be omitted for the server but it's required for the client. Certificate peer verification can be enabled/disabled on the Core side by changing `VerifyPeer` parameter in the configuration file. On the other hand, the SDL app library does not require the certificate from Core for the TLS handshake. However, Core performs its internal certificate validation before starting the actual TLS handshake. During internal validation, Core checks if the certificate is missing (or outdated/invalid) and if so, it initiates a PTU to obtain a new certificate from the Policy Server. If a valid certificate can't be obtained, Core does not start the TLS handshake and it notifies the app library that the protected service start has failed.

Reviewer: ```suggestion
The authentication is done using TLS handshake. The TLS handshake process is defined by TLS and is not part of the SDL protocol. The handshake is designed as a client-server communication which is configurable in the system settings. An application must take the role of a server where the system is the client. The client entity will initiate a TLS handshake with the corresponding security manager of the server. The client will do this only if the server was not authenticated before in the current transport connection. According to the TLS handshake process the peer certificate can be omitted for the server but it's required for the client. Certificate peer verification can be enabled/disabled on the Core side by changing `VerifyPeer` parameter in the configuration file. On the other hand, the SDL app library does not require the certificate from Core for the TLS handshake. However, Core performs its internal certificate validation before starting the actual TLS handshake. During internal validation, Core checks if the certificate is missing (or outdated/invalid) and if so, it initiates a PTU to obtain a new certificate from the Policy Server. If a valid certificate can't be obtained, Core does not start the TLS handshake and it notifies the app library that the protected service start has failed.
```

Author: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

The system can be configured to support one encryption method. The following methods are supported:

> - SSLv3

Reviewer: SSLv3 isn't supported on newer systems, I think it should either be removed or at least noted that it isn't supported on most systems.

Author: @jacobkeeler I moved this one to the bottom of the list and mentioned that it is not supported on newer systems. https://github.com/smartdevicelink/protocol_spec/pull/41/commits/14c425acd8be165cc36d355bfb658c98b336240a

  </tr>  
  <tr>
    <td>Encryption is disabled</td>
    <td>RPC Request doesn't need a protection</td>
    <td>SDL Core continues processing an RPC request</td>
  </tr>
  <tr>
    <td>Encryption is disabled</td>
    <td>RPC Notification doesn't need a protection</td>
    <td>SDL Core continues processing a notification</td>
  </tr>
  
  <tr>
    <td>Encryption is enabled</td>
    <td>RPC Request needs protection</td>
    <td>SDL Core rejects any unencrypted RPC requests with result code `ENCRYPTION_NEEDED` with the unencrypted response.</td>
  </tr>
  <tr>
    <td>Encryption is enabled</td>
    <td>RPC Notification needs protection</td>
    <td>SDL Core does not process an unencrypted notification</td>
  </tr>
  <tr>
    <td>Encryption is enabled</td>
    <td>RPC Notification doesn't need a protection</td>
    <td>SDL Core sends an unencrypted notification. </td>
  </tr>
  <tr>
    <td>Encryption is enabled</td>
    <td>RPC Request doesn't need a protection</td>
    <td>SDL Core continues processing an unencrypted RPC request if response is unencrypted. In addition, SDL Core shall continue processing an encrypted RPC request if response is encrypted.</td>
  </tr>

> </table>

Reviewer: ```suggestion
<table width="100%">
  <tr>
    <th width="30%">Service Encryption Status</th>
    <th width="10%">RPC Type</th>
    <th width="10%">Requires Protection</th>
    <th width="50%">Expected SDL Behavior</th>
  </tr>
  <tr>
    <td rowspan="4">Encryption is not established</td>
    <td rowspan="2">Request</td>
    <td>yes</td>
    <td>SDL Core rejects the request with result code `ENCRYPTION_NEEDED` (please see policy updates for which RPCs need protection).</td>
  </tr> 
  <tr>
    <td>no</td>
    <td>SDL Core continues processing the RPC request.</td>
  </tr>
  <tr>
    <td rowspan="2">Notification</td>
    <td>yes</td>
    <td>SDL Core does not send the notification.</td>
  </tr> 
  <tr>
    <td>no</td>
    <td>SDL Core sends the notification unencrypted.</td>
  </tr>
  
  <tr>
    <td rowspan="4">Encryption is established</td>
    <td rowspan="2">Request</td>
    <td>yes</td>
    <td>
      If unencrypted, SDL Core rejects the request with an unencrypted response and result code `ENCRYPTION_NEEDED`.<br>
      If encrypted, SDL Core continues processing the request and sends an encrypted response.
    </td>
  </tr>
  <tr>
    <td>no</td>
    <td>
      If unencrypted, SDL Core continues processing the request and sends an unencrypted response.<br>
      If encrypted, SDL Core continues processing the request and sends an encrypted response.
    </td>
  </tr>
  <tr>
    <td rowspan="2">Notification</td>
    <td>yes</td>
    <td>SDL Core sends the notification encrypted.</td>
  </tr>
  <tr>
    <td>no</td>
    <td>SDL Core sends the notification unencrypted.</td>
  </tr>

</table>
```

Some suggestions to clean up the table

Author: @jacobkeeler thank you, applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/2f371a70d67b8598b9cec923524479660b31024c

The system can be configured to support one encryption method. The following methods are supported:

- TLSv1
- TLSv1.1
- TLSv1.2
- DTLSv1

> - SSLv3 (isn't supported on a newer systems)

Reviewer: ```suggestion
- SSLv3 (not supported on most newer systems)
```

Author: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/efddba1d5059d7aa763a2502acb600a726535001

    <td rowspan="2">Request</td>
    <td>yes</td>
    <td>
      If unencrypted, SDL Core rejects the request with an unencrypted response and result code `ENCRYPTION_NEEDED`.<br>
      If encrypted, SDL Core continues processing the request and sends an encrypted response.
    </td>
  </tr>
  <tr>
    <td>no</td>
    <td>
      If unencrypted, SDL Core continues processing the request and sends an unencrypted response.<br>
      If encrypted, SDL Core continues processing the request and sends an encrypted response.
    </td>
  </tr>
  <tr>
    <td rowspan="2">Notification</td>
    <td>yes</td>
    <td>SDL Core sends the notification encrypted.</td>
  </tr>
  <tr>
    <td>no</td>
    <td>SDL Core sends the notification unencrypted.</td>
  </tr>

</table>

### 7.1 Authentication

The authentication is done using TLS handshake. The TLS handshake process is defined by TLS and is not part of the SDL protocol. 

The below diagram shows the sequence of how the TLS handshake exchanges certificates to compute the master secret.

![TLS Handshake activity diagram](https://user-images.githubusercontent.com/5848997/122258220-cb8b7100-ce9e-11eb-9b2a-a6194b0d1b68.png)

> Please see [SDL Overview Guides](https://github.com/smartdevicelink/sdl_overview_guides/tree/develop/docs/Protected%20Services/index.md) for more details.

Reviewer: ```suggestion
Please see [SDL Overview Guides](https://smartdevicelink.com/en/guides/pull_request/sdl-overview-guides/security/protected-services/) for more details.
```
One more link fix after the latest changes to the Overview Guides PR

Author: @jacobkeeler applied in https://github.com/smartdevicelink/protocol_spec/pull/41/commits/4823c337e15e4dcbf5e5918bec884f1cff192c06
